File: primetest.js
// === PROFESSIONAL COMPUTATIONAL VERIFICATION ===
// Target: 100 million even cases

const LIMIT = 200_000_000;          // Highest even N to test
const BAR_WIDTH = 40;                // Width of the ASCII progress bar
const PROGRESS_STEP = LIMIT / 100;   // Update bar every 0.01% (~10k cases)

// High-resolution timer
typeof process !== 'undefined' && process.hrtime && typeof require !== 'undefined';
const hrtime = () => {
  if (typeof process !== 'undefined' && process.hrtime) {
    const [s, ns] = process.hrtime();
    return s + ns / 1e9;
  } else {
    return performance.now() / 1000;
  }
};

// Binary search helper for lower bound
function lowerBound(arr, target) {
  let lo = 0, hi = arr.length;
  while (lo < hi) {
    const mid = lo + ((hi - lo) >> 1);
    if (arr[mid] < target) lo = mid + 1;
    else hi = mid;
  }
  return lo;
}

// Generate all primes up to LIMIT via sieve of Eratosthenes
const generatePrimes = (max) => {
  const isPrime = new Uint8Array(max + 1);
  isPrime.fill(1);
  isPrime[0] = isPrime[1] = 0;
  const root = Math.sqrt(max) | 0;
  for (let i = 2; i <= root; i++) {
    if (isPrime[i]) {
      for (let j = i * i; j <= max; j += i) isPrime[j] = 0;
    }
  }
  const primes = [];
  for (let i = 2; i <= max; i++) if (isPrime[i]) primes.push(i);
  return primes;
};

// Fast omega (distinct prime divisors) using trial division
const omega = (n) => {
  let cnt = 0;
  if ((n & 1) === 0) {
    cnt++;
    do { n >>= 1; } while ((n & 1) === 0);
  }
  for (let f = 3; f * f <= n; f += 2) {
    if (n % f === 0) {
      cnt++;
      do { n /= f; } while (n % f === 0);
    }
  }
  if (n > 1) cnt++;
  return cnt;
};

(async () => {
  console.log(`Starting verification up to N = ${LIMIT.toLocaleString()}`);
  const memBefore = typeof process !== 'undefined'
    ? process.memoryUsage().heapUsed
    : performance.memory?.usedJSHeapSize || 0;
  const t0 = hrtime();
  const primes = generatePrimes(LIMIT);
  const t1 = hrtime();
  console.log(`Primes generated: ${primes.length.toLocaleString()} in ${(t1 - t0).toFixed(2)}s`);

  let tested = 0, verified = 0, sPrime = 0, sSemi = 0;
  const failures = [];
  const tv0 = hrtime();

  for (let N = 6; N <= LIMIT; N += 2) {
    tested++;
    let ok = false;

    const startIdx = lowerBound(primes, N >>> 1);
    for (let i = startIdx; i < primes.length; i++) {
      const p = primes[i];
      if (p >= N) break;
      const w = omega(N - p);
      if (w <= 2) {
        verified++;
        w === 1 ? sPrime++ : sSemi++;
        ok = true;
        break;
      }
    }
    if (!ok) failures.push(N);

    if (tested % PROGRESS_STEP === 0) {
      const pct = tested / (LIMIT/2) * 100;
      const filled = Math.floor((pct/100) * BAR_WIDTH);
      const bar = '[' + '='.repeat(filled) + ' '.repeat(BAR_WIDTH - filled) + ']';
      const elapsed = hrtime() - tv0;
      // Clear console for environments that support it
      if (typeof console.clear === 'function') console.clear();
      console.log(`Progress ${pct.toFixed(2)}% ${bar}` +
                  ` ${tested.toLocaleString()}/${(LIMIT/2).toLocaleString()} evens` +
                  ` | ${(tested/elapsed).toFixed(0)} cases/s`);
    }
  }

  const tv1 = hrtime();
  console.log("\n=== RESULTS ===");
  console.log(`Range tested: 6 to ${LIMIT.toLocaleString()} (even only)`);
  console.log(`Total cases:    ${tested.toLocaleString()}`);
  console.log(`Verified:       ${verified.toLocaleString()}`);
  console.log(`Failures:       ${failures.length}`);
  console.log(`Success rate:   ${(100 * verified / tested).toFixed(6)}%`);
  console.log(`Verification time: ${(tv1 - tv0).toFixed(2)}s`);
  console.log(`Overall rate:      ${(tested / (tv1 - tv0)).toLocaleString()} cases/s`);
  console.log(`Memory used:       ${(((typeof process !== 'undefined'
    ? process.memoryUsage().heapUsed : performance.memory?.usedJSHeapSize || 0) - memBefore)
    /1024/1024).toFixed(2)} MB`);

  if (failures.length === 0) {
    console.log("\n🏆 All cases verified successfully.");
    console.log(`  ω(s)=1: ${sPrime.toLocaleString()} (${(100 * sPrime / verified).toFixed(2)}%)`);
    console.log(`  ω(s)=2: ${sSemi.toLocaleString()} (${(100 * sSemi / verified).toFixed(2)}%)`);
  } else {
    console.error(`\n❌ ${failures.length} counterexamples! First few: ${failures.slice(0, 10).join(', ')}`);
  }
})();

File: primetest.py
import time
import sys
import os
import bisect
from collections import defaultdict

# Optional psutil import for memory monitoring
try:
    import psutil
    HAS_PSUTIL = True
except ImportError:
    HAS_PSUTIL = False

def sieve_of_eratosthenes(limit):
    """Generate all primes up to limit using optimized sieve"""
    is_prime = bytearray([True] * (limit + 1))
    is_prime[0] = is_prime[1] = False
    
    sqrt_limit = int(limit**0.5) + 1
    for i in range(2, sqrt_limit):
        if is_prime[i]:
            for j in range(i*i, limit + 1, i):
                is_prime[j] = False
    
    return [i for i in range(2, limit + 1) if is_prime[i]]

def omega_optimized(n):
    """Fast omega (distinct prime divisors) using optimized trial division"""
    if n <= 1:
        return 0
    
    count = 0
    
    # Handle factor 2 efficiently with bit operations
    if n & 1 == 0:
        count += 1
        while n & 1 == 0:
            n >>= 1
    
    # Check odd divisors only
    f = 3
    while f * f <= n:
        if n % f == 0:
            count += 1
            while n % f == 0:
                n //= f
        f += 2
    
    # If n is still > 1, it's a prime factor
    if n > 1:
        count += 1
        
    return count

def clear_console():
    """Clear console if possible"""
    try:
        if os.name == 'nt':  # Windows
            os.system('cls')
        else:  # Unix/Linux/macOS
            os.system('clear')
    except:
        pass

def progress_bar_advanced(current, total, bar_width=40, prefix="Progress"):
    """Advanced progress bar with percentage"""
    percent = 100.0 * current / total
    filled_length = int(bar_width * current // total)
    bar = '[' + '=' * filled_length + ' ' * (bar_width - filled_length) + ']'
    
    return f"{prefix} {percent:.2f}% {bar} {current:,}/{total:,}"

def get_memory_usage_mb():
    """Get current memory usage in MB (requires psutil)"""
    if not HAS_PSUTIL:
        return 0
    try:
        process = psutil.Process(os.getpid())
        return process.memory_info().rss / 1024 / 1024
    except:
        return 0

def test_conjecture_optimized(max_n=200_000_000, progress_step=None):
    """Optimized test of the conjecture using JS-style improvements"""
    
    if progress_step is None:
        progress_step = max_n // 100  # Update every 1%
    
    print(f"=== PROFESSIONAL COMPUTATIONAL VERIFICATION ===")
    print(f"Target: {max_n:,} (testing {max_n//2:,} even cases)")
    
    # Memory before
    mem_before = get_memory_usage_mb()
    
    # Generate primes with timing
    print(f"Generating primes up to {max_n:,}...")
    t0 = time.perf_counter()
    primes = sieve_of_eratosthenes(max_n)
    t1 = time.perf_counter()
    print(f"Primes generated: {len(primes):,} in {t1-t0:.2f}s")
    
    # Initialize counters
    tested = 0
    verified = 0
    s_prime = 0  # Cases where ω(s) = 1
    s_semi = 0   # Cases where ω(s) = 2
    failures = []
    
    # Start verification timing
    tv0 = time.perf_counter()
    total_cases = max_n // 2 - 2  # Even numbers from 6 to max_n
    
    print(f"\nStarting verification...")
    
    for N in range(6, max_n + 1, 2):
        tested += 1
        found_decomposition = False
        
        # Binary search optimization: find first prime > N/2
        min_prime = N // 2 + 1
        start_idx = bisect.bisect_left(primes, min_prime)
        
        # Check primes starting from the binary search position
        for i in range(start_idx, len(primes)):
            p = primes[i]
            if p >= N:
                break
                
            s = N - p
            w = omega_optimized(s)
            
            if w <= 2:
                verified += 1
                if w == 1:
                    s_prime += 1
                else:  # w == 2 or w == 0
                    s_semi += 1
                found_decomposition = True
                break
        
        if not found_decomposition:
            failures.append(N)
            print(f"\n❌ FAILURE: N = {N}")
        
        # Progress reporting with advanced bar
        if tested % progress_step == 0:
            elapsed = time.perf_counter() - tv0
            rate = tested / elapsed if elapsed > 0 else 0
            
            # Clear console for cleaner output
            clear_console()
            
            progress_msg = progress_bar_advanced(tested, total_cases)
            print(progress_msg)
            print(f"Rate: {rate:.0f} cases/s | Failures: {len(failures)}")
            
            if len(failures) > 0:
                print(f"Recent failures: {failures[-5:]}")
    
    # Final timing
    tv1 = time.perf_counter()
    verification_time = tv1 - tv0
    
    # Memory after
    mem_after = get_memory_usage_mb()
    
    # Final progress bar
    clear_console()
    print(progress_bar_advanced(total_cases, total_cases))
    
    print(f"\n=== RESULTS ===")
    print(f"Range tested: 6 to {max_n:,} (even only)")
    print(f"Total cases:    {tested:,}")
    print(f"Verified:       {verified:,}")
    print(f"Failures:       {len(failures)}")
    print(f"Success rate:   {100.0 * verified / tested:.6f}%")
    print(f"Verification time: {verification_time:.2f}s")
    print(f"Overall rate:      {tested / verification_time:,.0f} cases/s")
    
    if HAS_PSUTIL:
        print(f"Memory used:       {mem_after - mem_before:.2f} MB")
    else:
        print(f"Memory monitoring: unavailable (install psutil for memory stats)")
    
    if len(failures) == 0:
        print(f"\n🏆 All cases verified successfully!")
        print(f"  ω(s)=1: {s_prime:,} ({100.0 * s_prime / verified:.2f}%)")
        print(f"  ω(s)=2: {s_semi:,} ({100.0 * s_semi / verified:.2f}%)")
    else:
        print(f"\n❌ {len(failures)} counterexamples!")
        print(f"First few: {failures[:10]}")
    
    return failures

def test_small_examples_optimized():
    """Test small examples with optimized omega function"""
    print("Testing small examples with optimized omega:")
    
    test_cases = [6, 8, 10, 12, 14, 16, 18, 20]
    
    for N in test_cases:
        print(f"\nN = {N}:")
        found = False
        
        for p in range(N//2 + 1, N):
            if is_prime_simple(p):
                s = N - p
                omega_s = omega_optimized(s)
                if omega_s <= 2:
                    print(f"  {N} = {p} + {s}, ω({s}) = {omega_s}")
                    found = True
                    break
        
        if not found:
            print(f"  No decomposition found!")

def is_prime_simple(n):
    """Simple primality test for small numbers"""
    if n < 2:
        return False
    if n == 2:
        return True
    if n % 2 == 0:
        return False
    for i in range(3, int(n**0.5) + 1, 2):
        if n % i == 0:
            return False
    return True

if __name__ == "__main__":
    # Test small examples first
    test_small_examples_optimized()
    
    print("\n" + "="*60)
    
    # Choose your test size:
    
    # Quick test (1 million)
    # test_conjecture_optimized(max_n=1_000_000, progress_step=10_000)
    
    # Medium test (10 million) 
    # test_conjecture_optimized(max_n=10_000_000, progress_step=100_000)
    
    # Large test (100 million)
    # test_conjecture_optimized(max_n=100_000_000, progress_step=1_000_000)
    
    # Full JS-equivalent test (200 million)
    test_conjecture_optimized(max_n=200_000_000, progress_step=2_000_000)